### PR TITLE
Add priority field to File, expose instance_id, enable url fields specification, pagination and more

### PR DESCRIPTION
**Pull request by 66eli77**

### CODE REVIEW

**Comment by rtibbles** [comment]
Any concerns on this before I merge it?

**Comment by aronasorman** [comment]
Sorry, I missed the discussion on why this field is necessary. Can you add inline comments on why we need this field?

**Comment by 66eli77** [comment]
seems like people think it's more convenient to be able to look up this field and decide which file should be used when you have say 3 video files in hi-res, mid-res, and low-res.

**Comment by aronasorman** [comment]
So if I understand this correctly, `priority` will be used if the user downloading videos didn't specify a specific preset. Is that correct?

**Comment by 66eli77** [comment]
Yeah, that's what I think if we already have preset, the default/priority can be created at the client side based on screen size.

**Comment by jamalex** [comment]
This is essentially a `sort_order` field on the `File` objects within a `ContentNode` -- will be used for displaying a dropdown of format options on the content renderer page, as appropriate, as well as (in the absence of custom logic like @66eli77 says, e.g. based on screen size) selecting the default format to display if multiple are available.

**Comment by jamalex** [comment]
(and for now, in the content export process, this can come directly from the sort order on the associated `FormatPreset` objects)

File: kolibri/content/migrations/0001_initial.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # -*- coding: utf-8 -*-
-# Generated by Django 1.9.1 on 2016-06-27 23:19
+# Generated by Django 1.9.1 on 2016-06-29 21:25
```
Comment (jamalex) [inline kolibri/content/migrations/0001_initial.py:2]: You're not deleting and re-generating migrations again, are you? :8ball:

Comment (jamalex) [inline kolibri/content/migrations/0001_initial.py:2]: Other than this, looks good to merge. Please revert this migration to the original and then add a new migration on top of the old one -- we need to get into that workflow or we're going to have headaches.

**Comment by aronasorman** [comment]
I see. It would be helpful then to have inline comments describing what it's used for, and how it's expected to be used (like does it sort by highest to lowest first?). Or maybe we should just add in a `Meta` attribute for it, like so:

```
class Meta:
    ordering = ["-priority"]
```

**Comment by aronasorman** [comment]
Just so it's explicit how Files are ordered from now on.

**Comment by 66eli77** [comment]
I did delete the migration and regen lol

**Comment by 66eli77** [comment]
since this app isn't published yet, delete migration won't be a problem right?

**Comment by jamalex** [comment]
> since this app isn't published yet, delete migration won't be a problem right?
In this case, since people aren't generating their own content DBs, it shouldn't be an issue in practice. If it were in the main DB, then it would cause problems for devs, since their DBs would think the migration had already been applied, and wouldn't apply it again, but a field would be missing.

So, in this case, not a blocker -- just a habit that we now need to start avoiding.

**Comment by jamalex** [comment]
>  And we should just add in a Meta attribute for it
That sounds like a good idea. But perhaps having it in ascending order, so it can just be directly written from `sort_order`. Semantically, we can think of priority:1 as "1st priority" rather than "low priority".

File: kolibri/content/urls.py [hunk 2]
```diff
@@ -43,9 +44,10 @@ class Meta:
         fields = ['parent']
 
     def title_description_filter(self, queryset, value):
+        # only return the first 30 results to avoid major slow down
         return queryset.filter(
             Q(title__icontains=value) | Q(description__icontains=value)
-        )
+        )[0:30]
```
Comment (aronasorman) [inline kolibri/content/urls.py]: You can just change the `[0:30]` to `[:30]`. It's clearer IMO.

File: kolibri/plugins/learn/views.py [hunk 3]
```diff
@@ -15,8 +15,10 @@ def get_context_data(self, **kwargs):
         channel_id = getattr(self.request, "channel_id", "dummy_db")
         with using_content_database(channel_id):
             context = super(LearnView, self).get_context_data(**kwargs)
-            topics_serializer = ContentNodeSerializer(get_top_level_topics(),
+
+            mcontext = {'request': self.request, 'channel_id': channel_id}
+            topics_serializer = ContentNodeSerializer(get_top_level_topics(), context=mcontext,
                                                       many=True)
-            topics_serializer.context["channel_id"] = channel_id
+            # topics_serializer.context["channel_id"] = channel_id
```
Comment (aronasorman) [inline kolibri/plugins/learn/views.py]: Delete commented out code.

File: kolibri/content/urls.py (hunk 2 above)
Comment (66eli77) [inline kolibri/content/urls.py]: in future, there will be loading the second 30 results, the third 30 results

Comment (rtibbles) [inline kolibri/content/urls.py]: I would suggest we use pagination to achieve this: http://www.django-rest-framework.org/api-guide/pagination/

**Comment by 66eli77** [comment]
@rtibbles  pagination added

File: kolibri/content/serializers.py [hunk 4]
```diff
@@ -84,52 +85,46 @@ class ContentNodeSerializer(serializers.ModelSerializer):
...
+
+        # enable dynamic fields specification!
+        if not isinstance(self.context['request'], WSGIRequest):
+            if 'request' in self.context and self.context['request'].query_params.get('fields'):
+                fields = self.context['request'].query_params.get('fields').split(',')
+                # Drop any fields that are not specified in the `fields` argument.
+                allowed = set(fields)
+                existing = set(self.fields.keys())
+                for field_name in existing - allowed:
+                    self.fields.pop(field_name)
```
Comment (jamalex) [inline kolibri/content/serializers.py:106]: I don't think `pop` works on tuples. Perhaps just do `self.fields = existing.intersection(allowed)`

Comment (66eli77) [inline kolibri/content/serializers.py:106]: self.fields is not a tuple. The data structure looks like this:

```
{'files': FileSerializer(many=True, read_only=True):
    url = DualLookuplinkedIdentityField(lookup_field_1='channelmetadata_channel_id', lookup_field_2='pk', view_name='file-detail')
    storage_url = SerializerMethodField()
    id = IntegerField(label='ID', read_only=True)
    priority = IntegerField(allow_null=True, required=False)
    checksum = CharField(allow_blank=True, max_length=400, required=False)
    available = BooleanField(required=False)
    file_size = IntegerField(allow_null=True, required=False)
    ...
}
```

I'm not sure what this data structure is. but pop works on it.
Also self.fields is different from fields, existing and allowed. 
self.fields = existing.intersection(allowed) will not work.

**Comment by jamalex** [comment]
This PR is superseded by #201 (everything here was merged in to https://github.com/learningequality/kolibri/tree/content_db_integration), so I'm closing it here.

**Pull request closed without merging**